core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -591,71 +609,73 @@ boolean putIfAbsentComputeIfPresent(K key, V value, Consumer<OakWriteBuffer> com
                 }
             }
 
-            if (inTheMiddleOfRebalance(c) || finalizeDeletion(c, lookUp)) {
+            if (inTheMiddleOfRebalance(c) || finalizeDeletion(c, ctx)) {
                 continue;
             }
-            /* If {@code lookUp == null} i.e., there is no entry with that key in Oak, the previous version
+            /* If there is no entry with that key in Oak, the previous version
             associated with the entry is INVALID_VERSION.
             Otherwise, there is an entry with the given key and its value is deleted (or at least in the process of
-            being deleted). After calling {@code deleteValueFinish(Chunk<K, V>, LookUp)}, the version of this entry
-            should be negative.
+            being deleted). After calling {@code deleteValueFinish(Chunk<K, V>, ThreadContext)}, the version of this
+            entry should be negative.
             Concurrent insertions, however, may cause the version to become valid again.
             Thus, to ensure that the operation behaves as if it itself unlinked the deleted value, we take the minus
-            the absolute value of the version written in lookUp (this way, if the version is already negative, it
+            the absolute value of the version written in ctx (this way, if the version is already negative, it
             remains negative).
              */
 
             // we come here when no key was found, which can be in 3 cases:
             // 1. no entry in the linked list at all
             // 2. entry in the linked list, but the value reference is INVALID_VALUE_REFERENCE
             // 3. entry in the linked list, the value referenced is marked as deleted
-            if (lookUp != null) {
-                lookUp.version = -Math.abs(lookUp.version);
-                assert lookUp.entryIndex > 0;
+            if (ctx.isKeyValid()) {
+                negateVersion(ctx.value);
+                assert ctx.entryIndex > 0;
             } else {
-                // lookUp is null so there was no such key found, while EntrySet allocates the entry
-                // (holding the key) the new lookUp is going to be returned to be used by EntrySet's
+                // There was no such key found, while EntrySet allocates the entry
+                // (holding the key) ctx is going to be updated to be used by EntrySet's
                 // subsequent requests to write value
-                lookUp = c.allocateEntryAndKey(key);
-                if (lookUp == null) {
+                c.allocateEntryAndKey(ctx, key);
+                if (!ctx.isKeyValid()) {
```

Comment:
It is a bit weird that "ctx.isKeyValid" has so many meanings... For example, just above (line 630), it means that key was found in previous lookUp(). Here, it means that the entry array (or chunk) was full and we need to rebalance. And someone, inexperienced, can say that we have already checked that key is valid, why to check again :)
Ideally, I would prefer some boolean being returned from allocateEntryAndKey() with false meaning no allocation happened, need to rebalance.

Reply:
Good point! I changed `EntrySet.allocateEntry()` to return a boolean and updated the documentation.